### PR TITLE
Add autograd support to TokenSwitch dispatch and combine

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #187385
* #187366
* __->__ #181314

`dispatch` and `combine` are adjoint operations: backward of dispatch calls
combine, and backward of combine calls dispatch. Single dispatch/combine
public API following `nn.Module`/`torch.matmul` conventions:

  - With `out=(out_tokens, out_weights, out_idx)`: writes to caller-supplied
    buffers and returns them; no autograd (efficient buffer reuse).
  - Without `out`: allocates buffers internally, returns them with full
    autograd support via `_DispatchAutograd` / `_CombineAutograd`.

Subclasses implement `_dispatch`/`_combine` as the raw buffer-writing
primitives and inherit both modes from the base class. `topk_weights`
receives no gradient (routing metadata at a different byte width).

Authored with Claude.

### CODE REVIEW

File: torch/distributed/_token_switch.py [hunk 1]
```diff
@@ -14,6 +15,79 @@ class Routing:
...
+        ts._combine(routing, expert_tokens, out_tokens)
+        ctx.ts = ts
+        ctx.routing = routing
+        ctx.expert_shape = expert_tokens.shape
+        ctx.expert_dtype = expert_tokens.dtype
+        ctx.top_k = routing.topk_idx.shape[1]
+        return out_tokens
+
+    @staticmethod
+    def backward(ctx: Any, grad_out_tokens: torch.Tensor) -> tuple:
```
Comment (Skylion007) [inline torch/distributed/_token_switch.py]: Please type the return Tuple. It has two Literal[None] and a torch.Tensor looks like.

File: torch/distributed/_token_switch.py [hunk 2]
```diff
@@ -14,6 +15,79 @@ class Routing:
...
+        ctx.tokens_shape = tokens.shape
+        return out_tokens, out_topk_weights, out_topk_idx
+
+    @staticmethod
+    def backward(
+        ctx: Any,
+        grad_out_tokens: torch.Tensor,
+        grad_out_topk_weights: torch.Tensor,
+        grad_out_topk_idx: torch.Tensor,
+    ) -> tuple:
```
Comment (Skylion007) [inline torch/distributed/_token_switch.py]: Type this fully too, Pyrefly doesn't like partially typed ones

File: torch/distributed/_token_switch.py (hunk 1 above)
Comment (kwen2501) [reply torch/distributed/_token_switch.py]: Done — typed it explicitly as `tuple[None, None, torch.Tensor, None, None]` (one entry per forward input).

File: torch/distributed/_token_switch.py (hunk 2 above)
Comment (kwen2501) [reply torch/distributed/_token_switch.py]: Done — `_CombineAutograd.backward` now returns `tuple[None, None, torch.Tensor]`. Pyrefly passes.

**Review by kapilsh — APPROVED** [review]
LGTM! added couple of comments to previous PR in the stack #178712 . To summarize:

- `_max_recv_tokens_per_rank` might need to exist on base `TokenSwitch` class
- tests currently only use topk=1, may be test with 2-3 values of topk for better coverage
